Support Python 3.5 and add functionality

DougBurke opened the pull request:
# Note

You can see the new documentation at https://deproject-test.readthedocs.io/en/ciao-411/index.html - which I try to keep up to date with this PR

# Summary

Bumped the version number to 0.2.0.

Add support for Python 3 and improve behavior when the pylab backend is selected (rather than ChIPS).

Support supplying an array of theta values (one per annulus) rather than assuming a single theta for all annuli.

Add `deproject_from_xflt` routine which loads in the radii and annuli values from the XFLT keywords used by the XSPEC `projct` model (and the PHA files), rather than the user having to manually create the Deproject instance with these values.

Add the ability to easily tie parameter values between a subset of shells - with the `tie_par` method - and to ensure that annuli that are tied together are fitted simultaneously. The `untie_par` method removes the tie.

Added a `guess` method that performs a "un-projected" fit, as suggested by XSPEC's `projct` model, to get the fit parameters to a physically-meaningful starting position for the deprojection. This **may** help speed up the deprojection and avoid local minima.

Added `conf` and `covar` routines that perform onion-skin style confidence or covariance error analysis.

The `fit`, `conf`, and `covar` routines all return an Astropy Table which contains columns containing the fit or error analysis results per annulus. Some columns are always present (`annulus`, `rlo_ang`, `rhi_ang`, `rlo_phys`, `rhi_phys`, as well as the fitted parameter values), but others depend on the command (the `fit` function returns information on the per-fit results, such as statistic, reduced statistic, success flag, whereas the error routines return parameter errors). These structures also contain a density column (and possibly errors). Some of these columns are Astropy Quantities (i.e. contain units); these are the radii and density columns.

Added `density_plot` and `par_plot` methods to plot the values as a function of radius (either in arcsec or kpc). The `fit_plot`, `conf_plot`, and `covar_plot` results plot up fit results, with the latter two displaying error bars.

Added `get_shells` and `get_radii` helper methods to extract information about the current set up. The fit, confidence, and covariance results can be retrieved with `get_fit_results`, `get_conf_results`, and `get_covar_results`, assuming the relevant function has been called.

Updated the documentation to reflect these additions and to fix some link rot (also http to https changes), and changed over to Read-The-Docs theme. Docstrings have been changed over to NumPy style.

Changed from distutils to setuptools.

Add a very-basic set of tests and Travis CI hook (using the standalone Sherpa installation).

# Notes

The theta support (i.e. allowing it to vary with annulus) could be enhanced to support it varying per dataset  (e.g. in the case of the multi-obsid case), but that requires more work.

This still is restricted to circular symmetry.

DougBurke committed:
Clean up of the repository

Replace hgignore with gitignore and remove an old Makefile.

taldcroft commented:
@DougBurke - awesome!  

Due to time constraints and lack of personal motivation I've basically abandoned this, but have just invited you to be a collaborator with push rights.  So have at it!

taldcroft commented:
Or if you feel like forking this or moving it elsewhere that's also fine by me.

DougBurke commented:
Let's see where this set of commits leads...

DougBurke committed:
Examples: update to latest capabilities

One of the major changes is the switch of the M87 example from
chi2gehrels to chi2xspecvar

Improved support for both Astropy and Crates, where appropriate.
Remove fit_m87_da.py script

DougBurke committed:
Remove conf routine added in #1

The patch set introduces a separate conf routine (and covar), so the
newly-added routine is removed before the new code is added.

DougBurke committed:
Add covar/conf (issue #1) and data access, improve plotting, Python 3

Add support for Python 3. Remove the fixed precision used for the scaling
values between shells (5 decimal places) as it caused poor fits for some
set ups. Improvements to the plot support (both ChIPS and Matplotlib,
although the use of ChIPS is no-longer recommended or guaranteed to work
well). The set_source method can now be used multiple times (previously
this was likely to error out), but support for changing the model mid-analysis
is still limited (e.g. changes to the redshift is not guaranteed to be
propogated properly).

Support for multiple-datasets per annulus is limited in this release.

The installation now requires Astropy (< 3.1 for use in CIAO 4.11 due to
conflicting NumPy requirements). The cosmocalc module has been removed
as calculation of the angular-diameter disatance now uses the Astropy
cosmology module.

Add new functionality, including

 - require arguments to be Astropy quantities for radii, theta values,
   and the angular-diameter distance (if set)
 - added the deproject_from_xflt routine which uses the XFLTxxxx keyword
   values in the input PHA files to set the annuli (radii and theta) in
   the same way that the XSPEC projct model does
 - allow per-sector theta values
 - add unsubtract, group, and ungroup methods to work on all annuli at
   once
 - allow model values to be tied between consecutive annuli when doing the
   deprojected (onion-skin) fit; this is limited to direct linking of the
   parameters and does not support mathematical expressions
 - add a guess method which is based on the suggestion from the XSPEC
   projct documentation: it does an "unprojected" fit to each shell and
   then adjusts the normalization based on the shell contributions
 - add conf and covar routines to estimate parameter errors using the
   onion-skin approach (extends #1 to include the covariance method, but
   is a different implementation); errors which can not be determined
   are replaced by NaN (rather than the None seen in the standard Sherpa
   structures)
 - the fit, covar, and conf methods now return structured information
   on the results (as an Astropy table), with the radii being returned
   as Astropy quantities (in both angular and physical formst)
 - add accessor methods for information, such as the fit or error results
   per shell and how the shells are grouped together (e.g. get_radii,
   get_shells, get_fit_results, get_conf_results)
 - support calculating errors on the density, based on the confidence or
   covariance results for the model normalization
 - add methods to plot parameter and density values as a function of
   radius (using angular or physical scales); these methods use the
   naming scheme of <quantity>_plot (e.g. density_plot, fit_plot) so
   as not to clash with the standard Sherpa plot methods (in particular
   plot_fit)
 - validity checks on input values (e.g. radii are monotonically
   increasing)
 - switch to NumPy-style docstrings
 - the load_pha method now accepts DataPHA objects as well as file names
 - add a test suite (limited functionality) and run on Travis-CI (based on
   astrofrog/example-travis-conda)

The code has also been re-arranged so that it is in the `deproject` package,
but this change is backwards compatible for simple use cases such as

    from deproject import Deproject

DougBurke committed:
Docs: update to reflect new and updated functionality, support RTD

Move to the Read-The-Docs theme and MathJaX for inline mathematics.

The Read-The-Docs build uses the standalone version of Sherpa provided as a
conda package for this build. This means that the documentation can not be
generated at build time, since there is no support for the required XSPEC
models in this version of CIAO. Note that the Sherpa version is not
versioned, but there is a restriction to python 3.5.

Reorganize the documentation so that there are multiple pages.

We have not yet integrarated the standalone and CIAO versions of the
Sherpa documentation (and may never do so). For now switch to the
standalone version for the "sherpa.astro.ui" layer commands (via the
reference links) since they are more up-to-date (but less detailed)
than the CIAO ahelp versions.

DougBurke committed:
general clean up for a PyPI release of version 0.2.0
